### PR TITLE
Pull Request for #15: MD5 changes by operating system

### PR DESCRIPTION
**Pull request by veli-m-andirin**
@santiagohermo, could you please have a look at the change here?

The only thing we did was imposing a stable sort rule to avoid MD5 changes coming from changes in ordering. For changes coming from end of line settings, we prefer to leave it to Git per https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1180351460. For changes in binaries, we prefer not to pursue them per https://github.com/gslab-econ/gslab_r/issues/15#issuecomment-1116841049.

### CODE REVIEW

File: issue15/check_data_types.R [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+df_scatters <- read.csv('issue15/scatters_data.csv')
+
+SaveData(df_scatters, key = "ssyk96", outfile = "issue15/df_scatters.csv",
+         logfile = "issue15/df_scatters_manifest.log")
+
+SaveData(df_scatters, key = "ssyk96", outfile = "issue15/df_scatters.RDS",
+         logfile = "issue15/df_scatters_manifest.log", appendlog = T)
+
+SaveData(df_scatters, key = "ssyk96", outfile = "issue15/df_scatters.dta",
+         logfile = "issue15/df_scatters_manifest.log", appendlog = T)
```
Comment (santiagohermo) [inline issue15/check_data_types.R:41]: When I ran this in my computer I find that hashes in the data file manifests change for these data only. 

The `csv` files do not change at all, and all binaries change. Both I believe are to be expected.

File: SaveData/R/SaveData.R [hunk 2]
```diff
@@ -97,16 +98,9 @@ SaveData <- function(df, key, outfile, logfile = NULL, appendlog = FALSE, sortby
...
-      args <- list(df)
-      i <- 2
-      while(i<length(key)+2) {
-        args[[i]] <- df[[(key[[i-1]])]]
-        i <- i + 1
-      }
       
       if (sortbykey) {
-        df <- do.call(arrange, args)  # sort by key values
+        df <- setorderv(df, key)  # sort by key values
```
Comment (santiagohermo) [inline SaveData/R/SaveData.R:103]: `setorderv` works by reference, so I think that the assignment is unnecessary. Doing 

```R
        setorderv(df, key)  # sort by key values
```
should work.

**Review by santiagohermo — CHANGES_REQUESTED** [review]
Here is a review @veli-m-andirin! Find a few comments below.

One general question: `df` is of whatever object type the user is using at the time he calls `SaveData`, right? I wonder whether transforming `df` into a `data.table` would make things more stable. Or maybe we should add tests that save the same data using different object types (`tibble`, `data.frame`, `data.table`)?

To transform the object  to a `data.table` you would need to add a line that reads
```R
setDT(df)
```

File: SaveData/R/SaveData.R (hunk 2 above)
Comment (veli-m-andirin) [reply SaveData/R/SaveData.R]: Thanks @santiagohermo! Implemented in f4c8cf0.

File: issue15/check_data_types.R (hunk 1 above)
Comment (veli-m-andirin) [reply issue15/check_data_types.R:42]: Do hashes for CSV files change as well? For binaries, all are definitely expected. For CSV files, the stable ordering might save us from some hash changes, but not from all.

**Comment by veli-m-andirin** [comment]
Thanks @santiagohermo for your comments! 

We were outputting `data.frame` objects (there is a line that coerces into data frame). In f4c8cf0 I added a line that coerces into data.table before outputting. Can you rerun the test scripts and see if you get changing hashes? If so, maybe you can modify that line and coerce into a  tibble, and rerun. Then I can do a final rerun, and if I see changing hashes as well, we give up our fight against changing hashes for now :)

**Comment by santiagohermo** [comment]
Thanks @veli-m-andirin! I re-ran everything after your latest updates. This is how the diff looks:

image

No hash changes! Only the binaries are different, as expected.

The only thing that remains in my opinion is to drop Line 157:
https://github.com/gslab-econ/gslab_r/blob/f4c8cf06a17f27e812882c8b37d6558a8d7e9729/SaveData/R/SaveData.R#L157

**Comment by veli-m-andirin** [comment]
Thanks @santiagohermo! I removed that line in 48312f3.

**Review by santiagohermo — APPROVED** [review]
Thanks @veli-m-andirin! I'm approving. I left one little extra comment only. Feel free to wrap up after you resolve it!

**Pull request merged**